blog/2022-07-07-LFX-Blog-Sasha.md

```diff
@@ -0,0 +1,119 @@
+
+[dcm]: https://mentorship.lfx.linuxfoundation.org/project/5c828028-f91c-4969-b4de-9efdb27bb869
+[lfx]: https://lfx.linuxfoundation.org/tools/mentorship/
+
+This blogpost summarizes my work at Tremor as a mentee and shows what could be done next.
+
+
+# About Tremor
+
+Tremor is an event processing system with the goal of allowing users to handle a high volume of messages by viewing them as a stream flowing between different nodes of a graph. Events go in and out of this graph thanks to connectors.
+
+The interactions between each node and the connector configuration are defined using the [`troy`] programming language.
+
+[`troy`]: https://www.tremor.rs/docs/edge/language/
+
+
+# Abstract
+
+My mentorship at Tremor was focused on building a connector for the [ClickHouse] database engine. More specifically, I was focused on the *sink* part, the one that allows data to flow out of the Tremor application.
+
+[ClickHouse] is a database management system designed to allow for real-time analysis of high volumes of non-aggregated data. Its initial goal was to power the Yandex.Metrica analytics platform.
+
+[ClickHouse]: https://clickhouse.com/
+
+
+# The ClickHouse connector
+
+The next subsections describe what I did during the mentorship.
+
+
+## Interacting with a ClickHouse database in Rust
+
+My first goal was to find a way to send requests to a ClickHouse node from a Rust program. It was a good start because it allowed me to experiment on the ClickHouse side without caring about what was happening in Tremor. I spent around three weeks playing on a separate repository named [`scrabsha/plays-with-clickhouse`] and getting extensive knowledge from Matthias about how Tremor works from the inside.
+
+[`scrabsha/plays-with-clickhouse`]: https://github.com/scrabsha/plays-with-clickhouse
+
+I found two Rust crates that could help us send requests to a ClickHouse database: [`clickhouse`] and [`clickhouse-rs`]. `clickhouse` is the first one I considered. I had to discard it because it was focused on concrete Rust types and needed to know *at compile time* what each datatype is composed of. The second crate was a bit more low-level and allowed us to do what we need.
+
+[`clickhouse`]: https://crates.io/crates/clickhouse
+[`clickhouse-rs`]: https://crates.io/crates/clickhouse-rs
+
+
+## Writing a super simple sink
+
+Once I got every ClickHouse detail right, I started playing with the Tremor side. There were multiple connectors already implemented in Tremor. I picked the simplest ones and started copying parts of it and quickly got something working.
+
+The only challenge I encountered is that Tremor defines its own `Value` type, representing a value whose type is not really known at runtime, and uses it *a lot*. It was a bit disconcerting doing such things in Rust, as I felt I was writing dynamically-typed code in a statically-typed language, but I managed to get through it.
```

Comment:
Do you mean known at compile time instead of runtime here?
```suggestion
The only challenge I encountered is that Tremor defines its own `Value` type, representing a value whose type is not really known at compile time, and uses it *a lot*. It was a bit disconcerting doing such things in Rust, as I felt I was writing dynamically-typed code in a statically-typed language, but I managed to get through it.
```